### DISCUSSION TITLE
How to update the UI from other threads with a ModelRc?

### DISCUSSION THREAD

**Discussion by voloder** [slint-ui/slint#5300] [Q&A]
Hello everyone, I am having a problem understanding how to update UI from other threads in Rust. From what I know I should use the `invoke_from_event_loop` function, but if I try to use it with an reference counted `VecModel`, then it complains about  `Rc<VecModel<DeviceData>> cannot be sent between threads safely`. 

Relevant snippet:
```rs
tokio::spawn(async move {
        let mut devices: Vec<Device> = Vec::new();
        let mut ble = ble::BleAdapter::new();
        ble.init().await.unwrap();

        let mut events = ble.scan().await.unwrap();
        while let Some(event) = events.next().await {
            println!("Event: {:?}", event);
            match event {
                CentralEvent::DeviceDiscovered(id) => {
                    let name = "Device".to_string();
                    let mac = id.to_string();
                    let device = Device::new(name, mac);
                    devices.push(device);

                    let device_datas: Vec<slint_modules::DeviceData> =
                        devices.into_iter().map(|device| device.into()).collect();
                    let model = Rc::new(VecModel::from(device_datas));

                    let handle_copy = ui_handle.clone();
                    invoke_from_event_loop(move || handle_copy.unwrap().set_devices(model.into()));
                }                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                _ => {}
            }
        }
    });
 ```
The example shown at https://archive.slint.dev/0.3.4/docs/rust/slint/fn.invoke_from_event_loop only deals with a constant numeric value. Any help would be appreciated.

**Comment by bjorn** [comment]
The trick is to not sent those `Rc` references to other threads, because that isn't safe. Send only the data the thread actually needs, let the thread return with whatever data it gives back and only then resume interaction with the UI. You can create a future to wait on the thread using [`slint::spawn_local`](https://releases.slint.dev/1.6.0/docs/rust/slint/fn.spawn_local.html) (also see the Tokio example there).

(Note that you have linked to ancient documentation, from when `spawn_local` did not exist yet.)

See also https://github.com/slint-ui/slint/discussions/4377#discussioncomment-8206790.

**Reply by voloder** [reply]
Oh, thanks for the response. Seems like when Googling you always end up on the old docs somehow.
